src/backend/distributed/executor/multi_task_tracker_executor.c

```diff
@@ -2821,44 +2824,79 @@ TrackerHashCleanupJob(HTAB *taskTrackerHash, Task *jobCleanupTask)
 	}
 
-	/* give task trackers time to finish their clean up jobs */
-	sleepInterval = Max(minimumSleepInterval, RemoteTaskCheckInterval * 2) * 1000L;
-	pg_usleep(sleepInterval);
+	/* record the time when we start waiting for cleanup jobs to be sent */
+	startTime = GetCurrentTimestamp();
 
-	/* walk over task trackers to which we sent clean up requests */
-	taskTrackerCell = NULL;
-	foreach(taskTrackerCell, taskTrackerList)
+	/*
+	 * Walk over task trackers to which we sent clean up requests. Perform
+	 * these checks until it times out.
+	 *
+	 * We want to determine timedOut flag after the loop start to make sure
+	 * we iterate one more time after time out occurs. This is necessary to report
+	 * warning messages for timed out cleanup jobs.
+	 */
+	remainingTaskTrackerList = taskTrackerList;
+	while (list_length(remainingTaskTrackerList) > 0 && !timedOut)
 	{
-		TaskTracker *taskTracker = (TaskTracker *) lfirst(taskTrackerCell);
-		int32 connectionId = taskTracker->connectionId;
-		const char *nodeName = taskTracker->workerName;
-		uint32 nodePort = taskTracker->workerPort;
+		List *activeTackTrackerList = remainingTaskTrackerList;
+		ListCell *activeTaskTrackerCell = NULL;
+		TimestampTz currentTime = 0;
 
-		ResultStatus resultStatus = MultiClientResultStatus(connectionId);
-		if (resultStatus == CLIENT_RESULT_READY)
+		remainingTaskTrackerList = NIL;
+
+		pg_usleep(statusCheckInterval);
+		currentTime = GetCurrentTimestamp();
+		timedOut = TimestampDifferenceExceeds(startTime, currentTime, timeoutDuration);
+
+		foreach(activeTaskTrackerCell, activeTackTrackerList)
 		{
-			QueryStatus queryStatus = MultiClientQueryStatus(connectionId);
-			if (queryStatus == CLIENT_QUERY_DONE)
+			TaskTracker *taskTracker = (TaskTracker *) lfirst(activeTaskTrackerCell);
+			int32 connectionId = taskTracker->connectionId;
+			const char *nodeName = taskTracker->workerName;
+			uint32 nodePort = taskTracker->workerPort;
+
+			ResultStatus resultStatus = MultiClientResultStatus(connectionId);
+			if (resultStatus == CLIENT_RESULT_READY)
 			{
-				ereport(DEBUG4, (errmsg("completed cleanup query for job " UINT64_FORMAT
-										" on node \"%s:%u\"", jobId, nodeName,
-										nodePort)));
+				QueryStatus queryStatus = MultiClientQueryStatus(connectionId);
+				if (queryStatus == CLIENT_QUERY_DONE)
+				{
+					ereport(DEBUG4, (errmsg("completed cleanup query for job "
+											UINT64_FORMAT, jobId)));
 
-				/* clear connection for future cleanup queries */
-				taskTracker->connectionBusy = false;
+					/* clear connection for future cleanup queries */
+					taskTracker->connectionBusy = false;
+				}
+				else if (timedOut)
+				{
+					ereport(WARNING, (errmsg("could not receive response for cleanup "
+											 "query status for job " UINT64_FORMAT " "
+											 "on node \"%s:%u\" with status %d", jobId,
+											 nodeName, nodePort, (int) queryStatus),
+									  errhint("manually clean job resources on node "
```

Comment:
weird, we always start in small caps when we do errmsg, but start with capitals in errrhints.